tutorial/nixify/1_nix_flake.md

```diff
@@ -3,8 +3,249 @@
+## Introduce Flake
+
+Let's start by creating a `flake.nix` file in the root directory. The first step after this will be to write the basic template for the flake, which consists of `inputs` and `outputs`. 
+```nix
+{
+  inputs = {
+
+  };
+  outputs = {self, ...}: {
+
+  };
+}
+
+```
+
+```nix
+{
+  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+  outputs = {self, nixpkgs}: {
+    
+  };
+}
+
+```
+
+```nix
+{
+  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+  outputs = {self, ...}@inputs: {
+    
+
+  };
+}
+
+```
+Syntactic sugar ^
+
+```nix
+{
+  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+  outputs = {self, nixpkgs}: 
+    let
+      supportedSystems = [ "x86_64-linux" "x86_64-darwin" "aarch64-linux" "aarch64-darwin" ];
+      forAllSystems = nixpkgs.lib.genAttrs supportedSystems;
+      pkgs = forAllSystems (system: nixpkgs.legacyPackages.${system});
+    in
+  {};
+}
+
+```
+
+```nix
+{
+  # Inside outputs' `let` block
+  overlay = final: prev: {
+    todo-app = self.callCabal2nix "todo-app" ./. { };
+  };
+}
+
+```
+Talk about final, prev and self, super ^ (although a preference, I would like to say why final, prev helps me understand this better)
+
+```nix
+{
+  # Inside outputs' `let` block
+  haskellPackages' = forAllSystems (system: pkgs.${system}.haskellPackages.extend overlay);
+}
+```
+
+```nix
+{
+  ...
+  outputs = {self, nixpkgs}:
+    let
+      ...
+    in
+  {
+    packages = forAllSystems ( system: {
+      inherit (haskellPackages'.${system}) todo-app;
+      default = haskellPackages'.${system}.todo-app;
+    });
+  };
+}
+
+```
+Talk about how `inherit` above is same as `todo-app=haskellPackages'.${system}.todo-app` and how that can be also built using `nix build .#todo-app`
+
+
+## DevShell support
+
+```nix
+{
+  # Inside outputs
+  devShells = forAllSystems ( system: {
+    default = ...
+    shell1 = ...
+    shell2 = ...
+  });
+}
+
+```
+```nix
+{
+  # Inside outputs
+  devShells = forAllSystems ( system: {
+    default = haskellPackages'.${system}.shellFor {
+      packages = p : [
+        p.todo-app
+      ];
+      buildInputs = with haskellPackages'.${system}; [
+        cabal-install
+        todo-app
+      ];
+    };
+  });
+}
+
+```
+Talk about why shellFor ^
+
+## Flakify postgres and postgREST
+
+```nix
+{
+  # Inside outputs
+  apps = forAllSystems ( system: {
+    app1 = ...
+    app2 = ...
+    default = ...
+  });
+}
+
+```
+```nix
+{
+  # Inside outputs
+  apps = forAllSystems ( system: {
+    hello = {
+      type = "app";
+      program = toString pkgs.${system}.hello;
+    }
+    default = self.apps.hello;
+  });
+}
+
+```
+Talk about `type` and the output of `nix run .#hello`
+```nix
+{
+  # Inside outputs
+  apps = forAllSystems ( system: {
+    postgres =
+      let
+        pgsql = pkgs.${system}.postgresql;
+      in
+      {
+        type = "app";
+        program = toString ( 
+          pkgs.${system}.writeShellScript "pg" ''
```

Comment:
I'd use `writeShellApplication` instead as it uses shellcheck automatically.

Reply:
I have changed it to `writeShellApplication`! shellcheck is great. I have also added more checks in the script in scenarios where the database is running and data directory already exists.